### PR TITLE
landing_checks: block commits with REPO- flag mismatch (bug 2032267)

### PR DESCRIPTION
**Pull request by shtrom**
We use the Lando `repo.name`, e.g., firefox-autoland. This won't match the name that the VCT hook uses, e.g., autoland. This means we will need to disable the equivalent HgMO check after this has landed.

 * landing_checks: add support for `repo_name` to `LandingChecks`
 * landing_checks: block commits with `REPO-` flag mismatch  (bug 2032267)
 * tests: add `ids` to `test_check_commit_message_invalid_message`
 * tests: DRY `HG_PATCH_AUTHOR_COMMIT_MESSAGE_TEMPLATE` in `landing_checks`
 * tests: port `repolocked` tests from v-c-t

### CODE REVIEW

File: src/lando/utils/landing_checks.py [hunk 1]
```diff
@@ -510,6 +513,13 @@ def next_diff(self, patch_helper: PatchHelper):
             )
             return
 
+        if match := REPO_FLAG_RE.search(firstline):
```
Comment (shtrom) [inline src/lando/utils/landing_checks.py]: Note: this is the simplest implementation, added to the existing `CommitMessagesCheck`. This means we won't be able to selectively disable only the `REPO-` check.

I think it should be sufficient for our purpose, though.

File: src/lando/utils/landing_checks.py [hunk 2]
```diff
@@ -510,6 +513,13 @@ def next_diff(self, patch_helper: PatchHelper):
...
+            if match["repo"] != self.repo_name:
+                self.commit_message_issues.append(
+                    f"Commit locked to another repo than {self.repo_name}: {commit_message}"
+                )
+                return
```
Comment (shtrom) [inline src/lando/utils/landing_checks.py:526]: Question: By `return`ing early here, and in other checks below, we only give a single error at a time per commit message. Should we instead not `return` and present all errors at once?

Comment (cgsheeh) [reply src/lando/utils/landing_checks.py:526]: If we can detect and warn about more issues that sounds like a good idea. Since we return a single string as the result of each check, it becomes a question of how the string is displayed in the UI.

I'd say this is acceptable for now, and in the future we can improve the checks so something more dynamic is returned from each warning/blocker.

File: src/lando/utils/tests/test_landing_checks.py [hunk 3]
```diff
@@ -215,7 +215,23 @@ def test_check_commit_message_valid_message(commit_message: str, error_message:
             ),
             "`git-format-patch` cruft should result in a failed check.",
         ),
+        (
```
Comment (cgsheeh) [inline src/lando/utils/tests/test_landing_checks.py:218]: Could we expand this to cover [more of the test cases from the hg hook](https://hg-edge.mozilla.org/hgcustom/version-control-tools/file/f21e253da4211b0b470fbced977be9522087b2d8/hghooks/tests/test-prevent-repolocked.t)?

File: src/lando/utils/landing_checks.py (hunk 1 above)
Comment (cgsheeh) [reply src/lando/utils/landing_checks.py]: Agreed, I think this is okay to move here. :)

**Review by cgsheeh — DISMISSED** [review]

File: src/lando/api/legacy/workers/automation_worker.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def run_job(self, job: AutomationJob) -> bool:
 
             if not self.skip_checks(job, new_commits) and repo.hooks_enabled:
                 patch_helpers = repo.scm.get_patch_helpers_for_commits(new_commits)
-                landing_checks = LandingChecks(job.requester_email)
+                landing_checks = LandingChecks(job.requester_email, repo.name)
```
Comment (shtrom) [inline src/lando/api/legacy/workers/automation_worker.py:105]: Using `repo.name`  here may be problematic, as the `repo.name` is the name in lando, e.g., `firefox-autoland`, which doesn't match the name that the VCT hook uses, e.g., `autoland`. This means either one or the other check will choke...

File: src/lando/utils/landing_checks.py (hunk 2 above)
Comment (shtrom) [reply src/lando/utils/landing_checks.py:526]: We return nothing here. We just add one error message with the commit message, and give up. We could add all the errors regarding every commit message in one go.

Given the way we currently render the errors, though, this would be trading off cognitive load for repeated partial fixes.

File: src/lando/utils/tests/test_landing_checks.py [hunk 5]
```diff
@@ -215,34 +215,199 @@ def test_check_commit_message_valid_message(commit_message: str, error_message:
...
+        # (
+        #     "Bug 1 - update file REPO-bar",
+        #     "Commit locked to a repo other than foo: ",
+        #     "Commits are disallowed when REPO-bar is in any commit in the push.",
+        # ),
+        # (
+        #     "Bug 1 - update file REPO-bar",
+        #     "",
+        #     "Commits locked to bar (with REPO-bar in the title) should be allowed on subdir/bar.",
+        # ),
```
Comment (shtrom) [inline src/lando/utils/tests/test_landing_checks.py:306]: Test case from v-c-t not ported.
```suggestion
```

File: src/lando/utils/tests/test_landing_checks.py [hunk 6]
```diff
@@ -215,34 +215,199 @@ def test_check_commit_message_valid_message(commit_message: str, error_message:
...
+        (
+            "Bug 1 - update file REPO-subdir/bar",
+            "Push contains commits intended to be locked to subdir/bar but the repo name is badly formatted. '/' is not allowed: ",
+            "Commits locked to subdir/bar (with REPO-subdir/bar in the title) should not be allowed on foo.",
+        ),
+        # (
+        #     "Bug 1 - update file REPO-subdir",
+        #     "",
+        #     "Commits locked to subdir (with REPO-subdir in the title) should not be allowed on subdir/bar.",
+        # ),
```
Comment (shtrom) [inline src/lando/utils/tests/test_landing_checks.py:321]: No ported.
```suggestion
```

File: src/lando/utils/tests/test_landing_checks.py [hunk 7]
```diff
@@ -215,34 +215,199 @@ def test_check_commit_message_valid_message(commit_message: str, error_message:
...
+        (
+            "Bug 1 - update file REPO-bar REPO-foo",
+            "",
+            "Commits locked to bar and foo should be allowed on foo.",
+        ),
+        # (
+        #     "Bug 1 - update file\n\nREPO-foo",
+        #     "",
+        #     "Commits locked to foo (with REPO-foo in the title) should be allowed on try.",
+        # ),
```
Comment (shtrom) [inline src/lando/utils/tests/test_landing_checks.py:336]: Not ported.
```suggestion
```

File: src/lando/utils/tests/test_landing_checks.py (hunk 3 above)
Comment (shtrom) [reply src/lando/utils/tests/test_landing_checks.py:218]: Done. That caught a few issues!

I added them all in the parametrise test cases, but some of them weren't relevant (as matching on HgMO repo paths). I left them commented out for review, with a suggestion to remove before merging.

File: src/lando/utils/landing_checks.py (hunk 2 above)
Comment (cgsheeh) [reply src/lando/utils/landing_checks.py:526]: Ah, I mis-read the code. In that case it does seem like we should let the hook continue looking for errors, but it's not a blocker.

**Review by cgsheeh — APPROVED** [review]
LGTM, please remove the commented out test cases before landing. :)

File: src/lando/utils/landing_checks.py (hunk 2 above)
Comment (shtrom) [reply src/lando/utils/landing_checks.py:526]: Yep. I filed a separate bug https://bugzilla.mozilla.org/show_bug.cgi?id=2033352

File: src/lando/api/legacy/workers/automation_worker.py (hunk 4 above)
Comment (shtrom) [reply src/lando/api/legacy/workers/automation_worker.py:105]: We'll disable the Hg-side hook https://bugzilla.mozilla.org/show_bug.cgi?id=2032267
